CONTRIBUTING.md

```diff
@@ -0,0 +1,437 @@
+    - Which features go in older stable versions is a tradeoff between how easy it is, how many mods are still using that version, and what the community is willing to contribute.
+- New features can be targeted at any supported version.
+    - Maintainers will take care of cherry-picking the feature to the other branches when applicable.
+    - In doubt, prefer PR'ing to newer stable versions.
+- Backporting PRs for older versions will generally be accepted, depending on how many changes were required.
+
+### Discouraged API design patterns
+
+- Avoid the old `Api.INSTANCE` pattern in favor of static methods.
+    - This pattern may be used if the backing implementation may be changed by another mod (such as the renderer api). Even then there should be static methods to access the standard api facilities.
+- Avoid unnecessary use of generics.
+    - Unless Vanilla Minecraft mandates the use of generics, or a good reason exists to use generics.
+    - In particular, use of self types should be avoided.
+- Avoid optionals in return values, fields and parameters.
+    - Where possible, you should prefer a `@Nullable T`.
+    - If vanilla exposes optionals in return types, then returning an optional is fine.
+- Avoid requiring the user to cast to a subtype if possible.
+    - Adding methods to vanilla types can be done via interface injection.
+
+### API design patterns to consider
+
+- Transitive Access Wideners (TAWs) should be used to expose access to private or protected members in vanilla classes.
+    - Most TAWs should go in the dedicated `fabric-transitive-access-wideners-v1` module.
+        - Remember to add the `transitive-` prefix, otherwise dependent mods will not see the access modifications.
+        - Some TAWs such as Block subclass constructors are automatically generated. Make sure that you don't edit the `.accesswidener` file in `src` directly. Rather edit the template file (`template.accesswidener`) and run the `gradlew generateAccessWidener` task to update the generated file.
+        - Large amounts of TAWs for a specific purpose can be included in another module, as is the case for the data generation API, for example.
+    - Do **not** expose TAWs for functions that take a `String` identifier.
+        - This makes it too easy to forget the mod ID namespace, so the identifier would often end up in the vanilla `minecraft` namespace.
+    - In general, keep the API guidelines in mind when deciding whether something should be a TAW.
+- Interface injection (i.e. making a minecraft class or interface extend a Fabric interface) should be considered over separate static helpers.
+    - Interface injection requires both a `fabric.mod.json` custom value to make it visible in Minecraft source code, and a mixin to actually implement the interface at runtime.
+    - Injected interfaces should have **no abstract methods**.
```

Comment:
Might be good to include the rationale here (that javac will throw a fit if it can't find the implementation of an itf method on a class iirc)

Reply:
Well, obviously we don't want to force modders to add methods beyond what vanilla requires. :P

Reply:
Well yeah, that too. But the javac thing applies when you're injecting into a final class or a class that no one will ever want to extend (LootTable)

Reply:
Should be ok now.